utilix/rundb.py

```diff
@@ -387,34 +388,64 @@ def delete_mc_document(self, document):
         return self._delete(url, data=doc)
 
-    # TODO need to be careful about versioning!!!
-    def download_file(self, filename, save_dir='./', force=False):
-        """Downloads file from GridFS"""
-        url = f'/files/{filename}'
-        os.makedirs(save_dir, exist_ok=True)
-        write_to = os.path.join(save_dir, filename)
-        if os.path.exists(write_to) and not force:
-            logger.debug(f"{filename} already exists at {write_to} and the 'force' flag is not set.")
-        else:
-            logger.debug(f"Downloading {filename} from gridfs...")
-            response = self._get(url)
-            with open(write_to, 'wb') as f:
-                f.write(response.content)
-            logger.debug(f'DONE. {filename} downloaded to {write_to}')
-        return write_to
-
-    def load_file(self, filename, save_dir=None, force=False):
-        if save_dir is None:
-            save_dir = os.path.join(os.environ.get("HOME"), '.gridfs_cache')
-        path = self.download_file(filename, save_dir=save_dir, force=force)
-        return io.read_file(path)
-
-    def upload_file(self, filepath):
+    # def download_file(self, filename, save_dir='./', force=False):
+    #     """Downloads file from GridFS"""
+    #     url = f'/files/{filename}'
+    #     os.makedirs(save_dir, exist_ok=True)
+    #     write_to = os.path.join(save_dir, filename)
+    #     if os.path.exists(write_to) and not force:
+    #         logger.debug(f"{filename} already exists at {write_to} and the 'force' flag is not set.")
+    #     else:
+    #         logger.debug(f"Downloading {filename} from gridfs...")
+    #         response = self._get(url)
+    #         with open(write_to, 'wb') as f:
+    #             f.write(response.content)
+    #         logger.debug(f'DONE. {filename} downloaded to {write_to}')
+    #     return write_to
+    #
+    # def load_file(self, filename, save_dir=None, force=False):
+    #     if save_dir is None:
+    #         save_dir = os.path.join(os.environ.get("HOME"), '.gridfs_cache')
+    #     path = self.download_file(filename, save_dir=save_dir, force=force)
+    #     return io.read_file(path)
+    #
+    # def upload_file(self, filepath):
+    #     with open(filepath, 'rb') as f:
+    #         fb = f.read()
+    #     filename = os.path.basename(filepath)
+    #     url = f'/files/{filename}'
+    #     return self._post(url, data=fb)
+
+    def count_documents(self, query: dict)->int:
+        """Perform collection.count_documents on the fs.files-collection using the query"""
+        """<URL MAGIC>"""
+        response = json.loads(self._get(url).text)
+        return response.get('results', 0)
+
+    def file_find(self, projection:dict)->dict:
+        """Perform collection.find(projection=projection) on the fs.files-collection"""
+        """<URL MAGIC>"""
+        response = json.loads(self._get(url).text)
+        return response.get('results', {})
+
+    def get_last_version(self, query: dict):
```

Comment:
This is effectively what the `download_file` method is doing, but it takes a filename instead of a dictionary query. I feel like a simple filename is more user friendly than a dict -- am I missing something though?

Reply:
Sure, we can do this. I had set it up to allow more complex queries, see https://github.com/XENONnT/utilix/pull/38#discussion_r600223954. Since this downloader has been working quite well without any issues for CMT/WFSim/straxen etc, we might not need more complex identifiers than solely the filename